p2p/src/main/kotlin/maru/p2p/P2PManager.kt

```diff
@@ -0,0 +1,130 @@
+import java.nio.file.Path
+import java.util.Optional
+import java.util.concurrent.TimeUnit
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.networking.p2p.libp2p.MultiaddrPeerAddress
+import tech.pegasys.teku.networking.p2p.libp2p.PeerAlreadyConnectedException
+import tech.pegasys.teku.networking.p2p.network.P2PNetwork
+import tech.pegasys.teku.networking.p2p.network.PeerAddress
+import tech.pegasys.teku.networking.p2p.network.config.GeneratingFilePrivateKeySource
+import tech.pegasys.teku.networking.p2p.peer.DisconnectReason
+import tech.pegasys.teku.networking.p2p.peer.NodeId
+import tech.pegasys.teku.networking.p2p.peer.Peer
+
+class P2PManager(
+  val privateKeyFile: Path,
+  val ipAddress: String,
+  val port: String,
+  val staticPeers: List<String>,
+) {
+  companion object {
+    private const val DEFAULT_RECONNECT_DELAY_MILLI_SECONDS = 5000L // TODO: Do we want to make this configurable?
+  }
+
+  private val log: Logger = LogManager.getLogger(this::class.java)
+  private val delayedExecutor = SafeFuture.delayedExecutor(DEFAULT_RECONNECT_DELAY_MILLI_SECONDS, TimeUnit.MILLISECONDS)
+  private val staticPeerMap = mutableMapOf<NodeId, MultiaddrPeerAddress>()
+
+  lateinit var p2pNetwork: P2PNetwork<Peer>
+
+  fun start() {
+    p2pNetwork = buildP2PNetwork()
+
+    p2pNetwork
+      .start()
+      ?.thenApply {
+        staticPeers.forEach { peer ->
+          p2pNetwork.createPeerAddress(peer)?.let { address -> addStaticPeer(address as MultiaddrPeerAddress) }
+        }
+      }?.get()
+  }
+
+  fun stop() {
+    p2pNetwork.stop()
+  }
+
+  private fun buildP2PNetwork(): P2PNetwork<Peer> {
+    val filePrivateKeySource = GeneratingFilePrivateKeySource(privateKeyFile.toString())
+    // TODO: reading/generating this key should be done early on, as it is needed for the validator as well
+    val privKeyBytes = filePrivateKeySource.privateKeyBytes
+    val privateKey = unmarshalPrivateKey(privKeyBytes.toArrayUnsafe())
+
+    return P2PNetworkFactory.build(
+      privateKey = privateKey,
+      ipAddress = ipAddress,
+      port = port,
+    )
+  }
+
+  fun addStaticPeer(peerAddress: MultiaddrPeerAddress) {
+    if (peerAddress.id == p2pNetwork.nodeId) { // Don't connect to self
+      return
+    }
+    synchronized(this) {
+      if (staticPeerMap.containsKey(peerAddress.id)) {
+        return
+      }
+      staticPeerMap.put(peerAddress.id, peerAddress) // This now works because staticPeers is mutable
+    }
+    maintainPersistentConnection(peerAddress)
+  }
+
+  fun removeStaticPeer(peerAddress: PeerAddress) {
+    if (staticPeerMap.remove(peerAddress.id) != null) {
+      p2pNetwork.getPeer(peerAddress.id).ifPresent { peer -> peer.disconnectImmediately(Optional.empty(), true) }
+    }
+  }
+
+  private fun maintainPersistentConnection(peerAddress: MultiaddrPeerAddress): SafeFuture<Void> {
```

Comment:
Let's stick to Unit instead of Void in the return type

Reply:
done